### PR TITLE
[v1.13] Add workaround for possible SetClusterProvider deadlock

### LINKED ISSUE #29952: [1.13] Deadlock in SetClusterProvider/libnetwork callbacks
https://jenkins.dockerproject.org/job/Docker-PRs-experimental/29248/console (from #29926) shows a failed test that hit a deadlock.

Daemon logs with traces: https://gist.github.com/tonistiigi/2ae3675901f18c3ccab7e6378aeaaf74

The issue seems to be that we are calling `SetClusterProvider()` while holding the cluster lock, the implementation needs a lock for a controller. Now the networking stack may call back to cluster `IsManager()` while holding the controller lock. `IsManager()` requires readlock for cluster.

This should be one of the cases solved with https://github.com/docker/docker/pull/28907 but that PR was not picked for `v1.13`.

We should evaluate the priority of this and see if there is a easy workaround that we could include in `v1.13`. One of the options could be to call `SetClusterProvider()` in a goroutine if it only works as a event dispatcher(cc @mavenugo to see if it would be safe).

There is also a possibility that recent changes in swarmkit (around graceful shutdown for example) have made this race more likely to appear.

@aaronlehmann @mavenugo @dongluochen

### PR DESCRIPTION
**Pull request by tonistiigi**
fixes #29952

@aaronlehmann @aboch 

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

### CODE REVIEW

File: daemon/daemon.go [hunk 1]
```diff
@@ -449,7 +449,9 @@ func (daemon *Daemon) registerLink(parent, child *container.Container, alias str
 // SetClusterProvider sets a component for querying the current cluster state.
 func (daemon *Daemon) SetClusterProvider(clusterProvider cluster.Provider) {
 	daemon.clusterProvider = clusterProvider
-	daemon.netController.SetClusterProvider(clusterProvider)
+	// call this in a goroutine to allow netcontroller handle this event async
```
Comment (thaJeztah) [inline daemon/daemon.go]: Can we add a TODO here if this is a workaround, and needs another approach?

Comment (tonistiigi) [inline daemon/daemon.go]: This issue has been fixed in master with https://github.com/docker/docker/pull/28907

Comment (thaJeztah) [inline daemon/daemon.go]: Ah, ignore then 👍

**Comment by aboch** [comment]
Looks good to me

**Review by aaronlehmann — APPROVED** [review]
LGTM

**Review by aboch — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
question: this fix reduces the probability of deadlock, but the deadlock can still happen, right?
If so, I think it is worth commenting.

Change itself is LGTM

**Comment by tonistiigi** [comment]
@AkihiroSuda The deadlock should not happen through `SetClusterProvider`. If you look at #29952 then there is one another trace that doesn't use `SetClusterProvider` but we think this it is impossible to hit it with real-world usage. Still, this is a workaround, the issue is fixed with #28907

**Comment by vieux** [comment]
SGTM

**Pull request merged**